### PR TITLE
Example/distributed embeddings training

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/ml/include/ml/optimisation/optimiser.hpp [hunk 1]
```diff
@@ -478,6 +480,18 @@ typename Optimiser<T>::SizeType Optimiser<T>::UpdateBatchSize(SizeType const &ba
   return updated_batch_size;
 }
 
+template <typename T>
+void Optimiser<T>::AddExternalGradientsToGraph(std::vector<TensorType> grads)
```
Comment (baykaner) [inline libs/ml/include/ml/optimisation/optimiser.hpp]: - does it matter that they're "external" - to me it makes sense if this function is called "AddGradientsToGraph".
- Add function comment (explain typical use case of injecting external gradients)

File: libs/ml/include/ml/ops/weights.hpp [hunk 2]
```diff
@@ -135,6 +135,11 @@ class Weights : public fetch::ml::ops::PlaceHolder<T>, public Trainable<T>
     return {};
   }
 
+  void AddExternalGradient(TensorType const &extern_grad)
```
Comment (baykaner) [inline libs/ml/include/ml/ops/weights.hpp]: why not just "AddToGradient"?
function comment

File: libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp [hunk 3]
```diff
@@ -48,29 +49,42 @@ class TrainingClient
...
+  virtual void     PrepareModel()      = 0;
+  virtual void     PrepareDataLoader() = 0;
+  virtual void     PrepareOptimiser()  = 0;
 
-private:
+protected:
   // Client's own graph and mutex to protect it's weights
-  fetch::ml::Graph<TensorType> g_;
-  std::mutex                   model_mutex_;
+  //  std::shared_ptr<fetch::ml::Graph<TensorType>> g_ptr_ =
```
Comment (baykaner) [inline libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp]: commented out code

File: libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp [hunk 4]
```diff
@@ -48,29 +49,42 @@ class TrainingClient
...
 
 public:
-  TrainingClient(std::string const &images, std::string const &labels, std::string const &id,
-                 SizeType batch_size, DataType learning_rate, float test_set_ratio,
+  TrainingClient(std::string const &id, SizeType batch_size, DataType learning_rate,
                  SizeType number_of_peers);
+  virtual ~TrainingClient() = default;
+
   void SetCoordinator(std::shared_ptr<Coordinator> coordinator_ptr);
   void MainLoop();
```
Comment (baykaner) [inline libs/ml/include/ml/distributed_learning/distributed_learning_client.hpp]: maybe lets just call this "Run()"?

File: libs/ml/include/ml/core/graph.hpp [hunk 5]
```diff
@@ -109,6 +109,8 @@ class Graph
   meta::IfIsNotGraphOrTrainable<TensorType, OperationType, void> AddTrainable(
       std::string const &name, std::shared_ptr<Node<TensorType>> node_ptr);
 
+  void AddExternalGradients(std::vector<TensorType> grads);
```
Comment (baykaner) [inline libs/ml/include/ml/core/graph.hpp]: AddGradients, or AddToGradients, are better names

File: libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+  this->inputs_names_ = {input_name, context_name};
+}
+
+template <class TensorType>
+void Word2VecClient<TensorType>::PrepareDataLoader()
+{
+
+  w2v_data_loader_ptr_ = std::make_shared<fetch::ml::dataloaders::GraphW2VLoader<DataType>>(
+      tp_.window_size, tp_.negative_sample_size, tp_.freq_thresh, tp_.max_word_count);
+  //  w2v_data_loader_ptr_->LoadVocab(vocab_file_);
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp]: commented out code left in

File: libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+ * Run model on test set to get test loss
+ * @param test_loss
+ */
+template <class TensorType>
+void Word2VecClient<TensorType>::Test(DataType &test_loss)
+{
+  // Implement loss mechanism
+  test_loss = static_cast<DataType>(0);
+
+  // Implement test frequency mechanism
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp]: left in todo?

File: libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+  using VectorTensorType = std::vector<TensorType>;
+
+public:
+  Word2VecClient(std::string const &id, TrainingParams<TensorType> const &tp,
+                 std::string const &vocab_file, SizeType batch_size, SizeType number_of_peers);
+  void PrepareModel() override;
+  void PrepareDataLoader() override;
+  void PrepareOptimiser() override;
+  void Test(DataType &test_loss) override;
+
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp:50]: I just want to clarify that this client is currently missing several important features that should be included in a future PR. For example these two:
- SetVocab(vocab_object) // overwrites existing vocab object
- SetVocabWord(String word, SizeType word_idx, SizeType count = numeric_max<SizeType>()) // update/re-assign a particular word to a particular index (and count if needed) - internally should remove previous assignment

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+#define LEARNING_RATE .001f
+#define TEST_SET_RATIO 0.03f
+#define NUMBER_OF_PEERS 2
+
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+using namespace fetch::ml;
+using namespace fetch::ml::dataloaders;
+
+using DataType         = double;
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: switch the datatype to fixed point by default. If the example does not work for fixed point we should investigate why.

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+using TensorType       = fetch::math::Tensor<DataType>;
+using VectorTensorType = std::vector<TensorType>;
+using SizeType         = typename TensorType::SizeType;
+
+// std::string ReadFile(std::string const &path)
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: left in comments

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    return 1;
+  }
+
+  std::string train_file = av[1];
+
+  std::shared_ptr<Coordinator> coordinator =
+      std::make_shared<Coordinator>(SYNCHRONIZATION_MODE, NUMBER_OF_ITERATIONS);
+
+  std::cout << "FETCH Distributed Word2vec Demo -- Asynchronous" << std::endl;
+  std::srand((unsigned int)std::time(nullptr));
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: don't use unsigned int - specify the size of the object: e.g. SizeType, uint32_t, etc.

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+  // calc the true starting learning rate
+  tp.starting_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.starting_learning_rate_per_sample;
+  tp.ending_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.ending_learning_rate_per_sample;
+  tp.learning_rate_param.starting_learning_rate = tp.starting_learning_rate;
+  tp.learning_rate_param.ending_learning_rate   = tp.ending_learning_rate;
+
+  //  GraphW2VLoader<DataType> data_loader(tp.window_size, tp.negative_sample_size, tp.freq_thresh,
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: left in comments

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+  // set up dataloader
+  /// DATA LOADING ///
+  //  std::string vocab_file = "/tmp/vocab.txt";
+  //  data_loader.BuildVocab({ReadFile(train_file)}, tp.min_count);
+  //  data_loader.SaveVocab(vocab_file);
+
+  std::vector<std::shared_ptr<TrainingClient<TensorType>>> clients(NUMBER_OF_CLIENTS);
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Instanciate NUMBER_OF_CLIENTS clients
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: typo instantiate

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    clients[i] = std::make_shared<Word2VecClient<TensorType>>(std::to_string(i), tp, train_file,
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: a future PR should replace the string IDs with something more sensible

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+
+    if (coordinator->GetMode() == CoordinatorMode::ASYNCHRONOUS)
+    {
+      break;
+    }
+
+    // Synchronize weights by giving all clients average of all client's weights
+    VectorTensorType new_weights = clients[0]->GetWeights();
+
+    // Sum all wights
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: typo

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+
+  std::vector<std::shared_ptr<TrainingClient<TensorType>>> clients(NUMBER_OF_CLIENTS);
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
+  {
+    // Instanciate NUMBER_OF_CLIENTS clients
+    clients[i] = std::make_shared<Word2VecClient<TensorType>>(std::to_string(i), tp, train_file,
+                                                              BATCH_SIZE, NUMBER_OF_PEERS);
+  }
+
+  for (unsigned int i(0); i < NUMBER_OF_CLIENTS; ++i)
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: replace all unsigned ints please

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+    // Wait for everyone to finish
+    for (auto &t : threads)
+    {
+      t.join();
+    }
+
+    if (coordinator->GetMode() == CoordinatorMode::ASYNCHRONOUS)
+    {
+      break;
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: shouldn't this be continue? Doesn't this break out of the for loop resulting in only 1 round?

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/examples/word2vec_distributed_learning/main.cpp (hunk 11 above)
Comment (MissingNO57) [reply libs/ml/examples/word2vec_distributed_learning/main.cpp]: This line shouldn't even exist, because we are using fetch:: random.

File: libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp (hunk 8 above)
Comment (MissingNO57) [reply libs/ml/examples/word2vec_distributed_learning/word2vec_client.hpp:50]: Yes, these features are missing and will be included in future PR.

File: libs/ml/examples/distributed_learning/main.cpp [hunk 18]
```diff
@@ -67,9 +68,10 @@ int main(int ac, char **av)
   for (SizeType i{0}; i < NUMBER_OF_CLIENTS; ++i)
   {
     // Instantiate NUMBER_OF_CLIENTS clients
-    clients[i] = std::make_shared<TrainingClient<TensorType>>(
+    clients[i] = std::make_shared<MNISTClient<TensorType>>(
```
Comment (baykaner) [inline libs/ml/examples/distributed_learning/main.cpp]: why make this mnist specific?

File: libs/ml/examples/distributed_learning/mnist_client.hpp [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/optimisation/adam_optimiser.hpp"
+#include <ml/distributed_learning/distributed_learning_client.hpp>
```
Comment (baykaner) [inline libs/ml/examples/distributed_learning/mnist_client.hpp]: style "" not <>

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+#include "ml/layers/skip_gram.hpp"
+#include "ml/ops/activation.hpp"
+#include "ml/ops/loss_functions.hpp"
+#include "ml/optimisation/sgd_optimiser.hpp"
+#include "model_saver.hpp"
+#include "word2vec_client.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cstddef>
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: i suspect some of these includes are surplus to requirements.
please find out which and remove them

File: libs/ml/examples/word2vec_distributed_learning/main.cpp [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#define NUMBER_OF_CLIENTS 3
```
Comment (baykaner) [inline libs/ml/examples/word2vec_distributed_learning/main.cpp]: replace these #defines with structs.
Specifically, one struct should encapsulate coordinator_params;
another struct should encapsulate experiment params.
Then pass the coordinator params object to the coordinator (this is a more general interface for the coordinator that we'll build on more later)

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/dataloaders/dataloader.hpp [hunk 22]
```diff
@@ -187,6 +187,8 @@ DataLoader<LabelType, InputType>::PrepareBatch(fetch::math::SizeType batch_size,
 template <typename LabelType, typename DataType>
 void DataLoader<LabelType, DataType>::SetMode(DataLoaderMode new_mode)
 {
+  if (mode_ == new_mode)
+    return;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/dataloader.hpp:192]: style - always use a curly brace

**Review by baykaner — APPROVED** [review]
